src/darwin/Framework/CHIP/MTRDeviceControllerStorageDelegate.h

```diff
@@ -111,6 +111,34 @@ MTR_NEWLY_AVAILABLE
     removeValueForKey:(NSString *)key
         securityLevel:(MTRStorageSecurityLevel)securityLevel
           sharingType:(MTRStorageSharingType)sharingType;
+
+@optional
+/**
+ * Return all keys and values stored, if any, for the provided controller in a
+ * dictionary. Returns nil if there is no stored value.
+ *
+ * securityLevel and dataType will always be the same for any given key value
+ * and are just present here to help locate the data if storage location is
+ * separated out by security level and data type.
+ *
+ * The set of classes that might be decoded by this function is available by
+ * calling MTRDeviceControllerStorageClasses().
+ */
+- (nullable NSDictionary<NSString *, id<NSSecureCoding>> *)valuesForController:(MTRDeviceController *)controller
+                                                                 securityLevel:(MTRStorageSecurityLevel)securityLevel
+                                                                   sharingType:(MTRStorageSharingType)sharingType;
+
+/**
+ * Store a list of keys and values in the form of a dictionary. Returns whether
+ * the store succeeded.
+ *
+ * securityLevel and dataType will always be the same for any given key value
+ * and are present here as a hint to how the value should be stored.
```

Comment:
I think the key part here is that the securityLevel and sharingType are the same across all the keys/values being stored in this store operation, right?

We need to very carefully document the expected behavior here if the stores fail partway through.  Is the callee expected to remove the already-stored things (roll back)?  Just leave things partially stored?  Something else?

Reply:
The other question is whether this should be the _only_ store API.  The "one value" case can just be represented as a dictionary with one entry...  We have not shipped these APIs yet, so can make that change if we think it's the right one.

Reply:
Yes same securityLevel and sharingType for each bulk operation.

I think the failure behavior that makes sense is that the client rolls back. Allowing partial failure would mean introducing a fine-grained return, perhaps a NSSet of keys that failed to store? But that seems a bit too much to me.

And I think this being the only API could work, as the complexity we'll add for the API consumer is at worst a loop.

Currently I have the bulk APIs separate as optional, so that a consumer could choose to take on the complexity if they wish, or only implement the single store/fetch.

Reply:
I've added that partial stores should just return NO.